bson/src/main/org/bson/codecs/pojo/PojoBuilderHelper.java

```diff
@@ -53,88 +44,105 @@
+        Map<String, TypeParameterMap> fieldTypeParameterMap = new HashMap<String, TypeParameterMap>();
+        Class<? super T> currentClass = clazz;
+        TypeData<?> classTypeData = getTypeData(currentClass.getGenericSuperclass(), currentClass);
 
-        TypeResolver resolver = new TypeResolver();
-        MemberResolver memberResolver = new MemberResolver(resolver);
-        ResolvedType resolved = resolver.resolve(clazz);
+        while (currentClass.getSuperclass() != null) {
+            annotations.addAll(asList(currentClass.getDeclaredAnnotations()));
 
-        ResolvedTypeWithMembers resolvedType = memberResolver.resolve(resolved,
-                new AnnotationConfiguration.StdConfiguration(AnnotationInclusion.INCLUDE_AND_INHERIT_IF_INHERITED), null);
+            List<String> genericTypeNames = new ArrayList<String>();
+            for (TypeVariable<? extends Class<? super T>> classTypeVariable : currentClass.getTypeParameters()) {
+                genericTypeNames.add(classTypeVariable.getName());
+            }
 
-        List<String> genericTypeNames = new ArrayList<String>();
-        for (TypeVariable<Class<T>> classTypeVariable : clazz.getTypeParameters()) {
-            genericTypeNames.add(classTypeVariable.getName());
-        }
+            for (Field field : currentClass.getDeclaredFields()) {
+                if (fieldNames.contains(field.getName()) || isTransient(field.getModifiers()) || isStatic(field.getModifiers())) {
+                    continue;
+                }
+                fieldNames.add(field.getName());
+                TypeParameterMap typeParameterMap = getTypeParameterMap(genericTypeNames, field);
+                fieldTypeParameterMap.put(field.getName(), typeParameterMap);
 
-        List<ResolvedField> resolvedFields = new ArrayList<ResolvedField>(asList(resolvedType.getMemberFields()));
+                FieldModelBuilder<?> fieldModelBuilder = getFieldBuilder(field, field.getType());
+                specializeFieldModelBuilder(fieldModelBuilder, typeParameterMap, classTypeData.getTypeParameters());
 
-        Map<String, TypeParameterMap> fieldTypeParameterMap = new HashMap<String, TypeParameterMap>();
-        for (final ResolvedField resolvedField : resolvedFields) {
-            if (resolvedField.isTransient()) {
-                continue;
+                classModelBuilder.addField(fieldModelBuilder);
+                field.setAccessible(true);
             }
-            String name = resolvedField.getName();
-            fieldTypeParameterMap.put(name, getTypeParameterMap(genericTypeNames, resolvedField));
-            classModelBuilder.addField(getFieldBuilder(resolvedField.getRawMember(), resolvedField.getType().getErasedType(),
-                    resolvedField.getType()));
-            resolvedField.getRawMember().setAccessible(true);
+
+            classTypeData = getTypeData(currentClass.getGenericSuperclass(), currentClass);
```

Comment:
Yeah, it is odd, we don't need to specialise at the top level, only for sub classes.  When specialising subclasses we have to use the parent classes type information.   I'll look to see if I can make that more obvious.